src/Data/Primitive/Vec.hs

```diff
@@ -83,6 +94,37 @@ import GHC.Word
 --
 data Vec (n :: Nat) a = Vec ByteArray#
 
+class Vectoring vector a | vector -> a where
+    type IndexType vector :: Data.Kind.Type
+    vecIndex :: vector -> IndexType vector -> a
+    vecWrite :: vector -> IndexType vector -> a -> vector
+    vecEmpty :: vector
+
+instance (KnownNat n, Prim a) => Vectoring (Vec n a) a where
+    type IndexType (Vec n a) = Int
+    vecIndex (Vec ba#) i@(I# iu#) = let
+        n :: Int
+        n = fromIntegral $ natVal $ Proxy @n
+        in if i >= 0 && i < n then indexByteArray# ba# iu# else error ("index " <> show i <> " out of range in Vec of size " <> show n)
+    vecWrite vec@(Vec ba#) i@(I# iu#) v = runST $ do
+        let n :: Int
+            n = fromIntegral $ natVal $ Proxy @n
+        mba <- newByteArray (n * sizeOf (undefined :: a))
+        let new_vs = zipWith (\i' v' -> if i' == i then v else v') [0..n-1] (listOfVec vec)
+        zipWithM_ (writeByteArray mba) [0..n-1] new_vs
```

Comment:
`copyByteArray` followed by `writeByteArray`, going via lists is obscene!

Reply:
But how to obtain a mutable reference to a byte array after it has been frozen?

Reply:
*cough* [unsafeCoerce#](https://hackage.haskell.org/package/primitive-0.7.3.0/docs/src/Data.Primitive.ByteArray.html#unsafeFreezeByteArray) *cough*